### DISCUSSION TITLE
Using API keys for Produce

### DISCUSSION THREAD

**Discussion by Alarson93** [fastlane/fastlane#17889] [Ideas] [answered]
Produce via API Key is not supported according to the [documentation](https://docs.fastlane.tools/app-store-connect-api/).

Apple will be requiring 2FA for all accounts in February and there have been [related threads](https://github.com/fastlane/fastlane/discussions/17655) created on here.

Not having produce won't be game breaking, but it will create manual steps. I'd like to know if support for Produce using the API Key is currently being worked on.
If so - is there an estimate on when it will be ready?
If not - why? Is there an API limitation that prevents Produce from functioning?

**Comment by dinsen** [comment, marked as answer]
According to this [documentation](https://developer.apple.com/documentation/appstoreconnectapi/apps) it's not possible to use the App Store Connect API to create new apps in App Store Connect.

**Reply by tmaly1980** [reply]
So... Then our only viable options are to either run produce with 2FA or to create the app listing manually (as we would with Google)? Will that need to be done before running gym, or can the app be built and ready to upload to testflight, at which point someone can manually create the listing and then run pilot (similar to how android is done) to upload the binary/metadata?

**Reply by dinsen** [reply]
I have not had this case before. So I don't know if you can run gym before the app is created. 
But I will like to know if you find the answer 😄

**Reply by rogerluan** [reply]
My guess is that you have to create it first 😅

**Reply by max-ott** [reply]
App record always needs to be created first as you will otherwise get errors from `Transporter` when trying to upload the app (`Record not found` or similar)

**Reply by wow-such-amazing** [reply]
https://developer.apple.com/documentation/appstoreconnectapi/register_a_new_bundle_id

This endpoint should partly work for what produce does, no? I mean creating an app in DevCenter 🤔

**Reply by max-ott** [reply]
`produce` creates the app on ADC and ASC. Endpoint is for ADC only.

**Reply by wow-such-amazing** [reply]
Yup, that's what I thought. I hope they will implement and endpoint for ASC soon 🤞 

Do you know if there are any plans to implement support in Fastlane only for ADC for now?

**Reply by max-ott** [reply]
I'm not aware of any pull requests or similar. It would add additional complexity and friction so I doubt there will be any initiative from our side.

**Reply by wow-such-amazing** [reply]
Makes sense. Thank you for you replies 🙂

**Comment by PeterMorrison250** [comment]
I was wondering if there have been any further discussions about this topic? If trying to automate the creation of the app on App Store Connect, is the only way forward with this remaining using the 2FA approach, instead of the Apple API keys?

**Comment by jakerenshawballys** [comment]
Is there any update on this? @iBotPeaches @matthewellis @KrauseFx 

Would be great to align produce with the other fastlane commands for automation purposes.

If not, what are the limitations still?

**Reply by iBotPeaches** [reply]
No update I'm aware of. I'm unsure if there is any technical limitation. Process limitation is time and money. If you'd like to fund it, we can probably remove one of the limitations.

**Reply by jakerenshawballys** [reply]
I did see this discussion as well - https://github.com/fastlane/fastlane/issues/18157#issuecomment-5240864777
(apologies on the tagging in multiple places btw)

**Reply by jakerenshawballys** [reply]
but thank you for the original answer, I don't offer funding myself but I do hope someone is able to remove the blockades you currently have in place 🙏

**Reply by jakerenshawballys** [reply]
Good afternoon @iBotPeaches, hope you're well 🙂

I've taken it upon myself to investigate the issue and I've raised a PR that addresses some of the problems of fastlane produce here if that's ok - 
https://github.com/fastlane/fastlane/pull/30158

Please let me know your thoughts or any changes I can make to hopefully get this into the main fastlane repo

Thanks!

CC @revolter @hjanuschka @giginet @KrauseFx @AliSoftware
